p2pserver/protocols/subnet/subnet.go

```diff
@@ -0,0 +1,359 @@
+
+func (self *SubNet) OnMembersRequest(ctx *p2p.Context, msg *types.SubnetMembersRequest) {
+	sender := ctx.Sender()
+
+	peerAddr := sender.Info.RemoteListenAddress()
+	if !self.checkAuthority(peerAddr, msg) {
+		self.logger.Info("[subnet] check authority for members request failed, peer: %s", peerAddr)
+		return
+	}
+
+	self.lock.Lock()
+	members := make([]types.MemberInfo, 0, len(self.members))
+
+	for addr, status := range self.members {
+		members = append(members, types.MemberInfo{PubKey: status.PubKey, Addr: addr})
+	}
+
+	//update self.members
+	if !msg.FromSeed() && self.gov.IsGovNode(msg.PubKey) && self.members[peerAddr] == nil {
+		self.members[peerAddr] = &MemberStatus{
+			PubKey: vconfig.PubkeyID(msg.PubKey),
+			Alive:  time.Now(),
+		}
+	}
+	self.lock.Unlock()
+
+	reply := &types.SubnetMembers{Members: members}
+	self.logger.Debugf("[subnet], send members to peer %s, value: %s", sender.Info.Id.ToHexString(), reply)
+	ctx.Network().SendTo(sender.GetID(), reply)
+}
+
+func (self *SubNet) OnMembersResponse(ctx *p2p.Context, msg *types.SubnetMembers) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	self.logger.Debugf("[subnet], receive members: %s ", msg.String())
+
+	listen := ctx.Sender().Info.RemoteListenAddress()
+	if self.connected[listen] == nil {
+		self.logger.Info("[subnet] receive members response from unkown node: %s", listen)
+		return
+	}
+
+	for _, info := range msg.Members {
+		if self.members[info.Addr] == nil {
+			self.members[info.Addr] = &MemberStatus{
+				PubKey: info.PubKey,
+				Alive:  time.Now(),
+			}
+		}
+	}
+
+	self.unparker.Unpark()
+}
+
+func (self *SubNet) getUnconnectedGovNode() []string {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	var addrs []string
+	for addr := range self.members {
+		if self.connected[addr] == nil {
+			addrs = append(addrs, addr)
+		}
+	}
+
+	return addrs
+}
+
+func (self *SubNet) newMembersRequest() *types.SubnetMembersRequest {
+	var request *types.SubnetMembersRequest
+	if self.IsSeedNode() {
+		request = types.NewMembersRequestFromSeed()
+	} else if self.acct != nil && self.gov.IsGovNode(self.acct.PublicKey) {
+		var err error
+		request, err = types.NewMembersRequest(self.acct)
+		if err != nil {
+			return nil
+		}
+	}
+
+	return request
+}
+
+func (self *SubNet) sendMembersRequestToRandNodes(net p2p.P2P) {
+	request := self.newMembersRequest()
+	if request == nil {
+		return
+	}
+
+	count := 0
+	peerIds := make([]common.PeerId, 0, MaxMemberRequests)
+	self.lock.RLock()
+	// note map iteration is randomized
+	for _, p := range self.connected {
+		count += 1
+		peerIds = append(peerIds, p.Id)
+		if count == MaxMemberRequests {
+			break
+		}
+	}
+
+	self.lock.RUnlock()
+	for _, peerId := range peerIds {
+		net.SendTo(peerId, request)
+	}
+}
+
+func (self *SubNet) sendMembersRequest(net p2p.P2P, peer common.PeerId) {
+	request := self.newMembersRequest()
+	if request == nil {
+		return
+	}
+
+	net.SendTo(peer, request)
+}
+
+func (self *SubNet) cleanStaleGovNode() {
```

Comment:
from my understanding, subnet is full-connected.
Should GovNode be cleaned when it is not gov node any more?
reconnect-service or some other keep-alive service should maintain their connections by pub-key list?
when sub-net service started, if current node is gov-node, it should try to discover all other gov-node and connect with them?

Reply:
retired gov node is disconnected and cleaned now. 
reconnect-service will fail after some attempt, since it ip is not in remote peer's reserve list.
discover logic is in subnet.go by sending members request message periodically.